homeassistant/components/guardian/__init__.py

```diff
@@ -89,33 +136,135 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     if unload_ok:
         hass.data[DOMAIN][DATA_CLIENT].pop(entry.entry_id)
         hass.data[DOMAIN][DATA_COORDINATOR].pop(entry.entry_id)
+        hass.data[DOMAIN][DATA_LAST_SENSOR_PAIR_DUMP].pop(entry.entry_id)
+        for async_unsub_dispatcher_connect in hass.data[DOMAIN][
+            DATA_UNSUB_DISPATCHER_CONNECTS
+        ][entry.entry_id]:
+            async_unsub_dispatcher_connect()
+        hass.data[DOMAIN][DATA_UNSUB_DISPATCHER_CONNECTS].pop(entry.entry_id)
 
     return unload_ok
 
 
-class GuardianEntity(Entity):
-    """Define a base Guardian entity."""
+class PairedSensorManager:
+    """Define an object that manages the addition/removal of paired sensors."""
 
     def __init__(
         self,
+        hass: HomeAssistant,
         entry: ConfigEntry,
         client: Client,
-        coordinators: Dict[str, DataUpdateCoordinator],
-        kind: str,
-        name: str,
-        device_class: str,
-        icon: str,
+        api_lock: asyncio.Lock,
+    ) -> None:
+        """Initialize."""
+        self._api_lock = api_lock
+        self._client = client
+        self._entry = entry
+        self._hass = hass
+        self._paired_uids = set()
+
+    async def async_init(self) -> None:
+        """Perform post-creation initialization."""
+        for signal_template, target in [
+            (SIGNAL_ADD_PAIRED_SENSOR, self.async_pair_sensor),
+            (SIGNAL_REMOVE_PAIRED_SENSOR, self.async_unpair_sensor),
+        ]:
+            async_register_dispatcher_connect(
```

Comment:
It feels weird that this registers things with a global teardown. I would expect that to be done inside this class.

Reply:
This isn't clear. Are you saying I shouldn't use the dispatcher here (and should instead of entities call into this class) when adding/removing things?

Reply:
`PairedSensorManager` owns these dispatch subscriptions. So it should also be responsible for unsubscribing when this class is teared down. By adding the subscriptions to a global object seems weird because then it's not clear what resources this class manages.

Reply:
Check out 68fe3cf – the `PairedSensorManager` no longer establishes any dispatcher connections directly.